---
title: "BME 4550 - Assignment 9"
author: "Anthony (Tony) DeNovio"
date: "November 7, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dagdata)
library(Biobase) 
library(GSE5859) 
library(qvalue)
library(genefilter)
library(GSE5859Subset) 
library(RColorBrewer)
library(rafalib)
data(GSE5859Subset)
data(admissions)
```

#\  

####**On my honor as a student, I have neither given nor received unauthorized aid on this assignment.**

####-*Anthony DeNovio*

#\

####Note: I left all r code blocks as eval = FALSE to speed up knitting time.

#\  

####All questions have been **bolded** and answers have been *italicized*.

#\  

#Batch Effects Exercise List 1:

###**1. Compute the proportion of men who were accepted**

```{r, eval = FALSE}
index = which(admissions$Gender==1) 
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100) 
applied = sum(admissions$Number[index]) 
accepted/applied
```

####*0.4451951.*

###**What is the proportion of women that were accepted?** 

```{r, eval = FALSE}
index = which(admissions$Gender==0) 
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100) 
applied = sum(admissions$Number[index]) 
accepted/applied
```

####*0.3033351.*

#\ 

###**2. Now that we have observed different acceptance rates between genders, test for the significance of this result. If you perform an independence test, what is the p-value?**

```{r, eval = FALSE}
index = admissions$Gender==1 
men = admissions[index,] 
women = admissions[!index,] 
menYes = sum(men$Number*men$Percent/100) 
menNo = sum(men$Number*(1-men$Percent/100)) 
womenYes = sum(women$Number*women$Percent/100) 
womenNo = sum(women$Number*(1-women$Percent/100)) 
tab = matrix(c(menYes,womenYes,menNo,womenNo),2,2) 
print(chisq.test(tab)$p.val)
```

####*p-value = 9.139492e-22.*

#\ 

###**3. We can quantify how “hard” a major is by using the percent of students that were accepted. Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major?** 
```{r, eval = FALSE}
major = admissions[1:6,1]
men = admissions[1:6,]
women =admissions[7:12,]
H = (men$Number*men$Percent/100 + women$Number*women$Percent/100) / (men$Number+women$Number)
major[which.min(H)]
```

####*Major F.*

#\ 

###**4. What proportion is accepted for this major?** 

```{r, eval = FALSE}
min(H)
view(admissions)
```

####*0.06477591.*

#\ 

###**5. For men, what is the correlation between the number of applications across majors and H?**

```{r, eval = FALSE}
cor(men$Number,H)
```

####*0.7647567.*

#\ 

###**6. For women, what is the correlation between the number of applications across majors and H?**

```{r, eval = FALSE}
cor(H,women$Number)
```

####*-0.6743393.*

#\ 

###**7. Given the answers to the above, which best explains the differences in admission percentages when we combine majors?** 

####A) We made a coding mistake when computing the overall admissions percentages.

####B)There were more total number of women applications which made the denominator much bigger. 

####C) There is confounding between gender and preference for “hard” majors: females are more likely to apply to harder majors. 

####D)The sample size for the individual majors was not large enough to draw the correct conclusion.

####*C.*

#\ 

#Batch Effects Exercise List 2:

###**1. Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here. You can extract the year from each date like this:**

```{r, eval = FALSE}
geneExpression = exprs(e) 
sampleInfo = pData(e)
year = format(sampleInfo$date,"%y")
```

#\ 

###**2. For how many of these years do we have more than one ethnicity represented?**

```{r, eval = FALSE}
year = format(sampleInfo$date,"%y")
tab = table(year,sampleInfo$ethnicity)
x=rowSums(tab!=0)
sum(x>=2)
```

####*Two.*

#\ 

###**3. For what proportion of these month.year values do we have more than one ethnicity represented?**

```{r, eval = FALSE}
month.year = format(sampleInfo$date,"%m%y")
tab = table(month.year,sampleInfo$ethnicity)
print(tab)
x = rowSums(tab!=0)
mean(x>=2)
```

####*0.04762*

#\ 

###**4. Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values < 0.05?** 

```{r, eval = FALSE}
year = factor( format(sampleInfo$date,"%y") )
ind = which(year%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[ind])
pval = rowttests(geneExpression[ ,ind], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

####*4308 genes.*

#\ 

###**5. What is the estimate of pi0 provided by qvalue:** 

```{r, eval = FALSE}
qval$pi0
```

####*pi0 ~ 0.3628642*

#\ 

###**6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004. Then use the q value package to obtain q-values for each gene. How many genes have q-values less than 0.05?**

```{r, eval = FALSE}
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

####*2463 genes.*

#\ 

###**7. Now we are going to compare ethnicities as was done in the original publication in which these data were first presented. Use the qvalue function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values. How many genes have q-values < 0.05?** 

```{r, eval = FALSE}
ethnicity = factor( sampleInfo$ethnicity)
index = which(ethnicity%in% c("CEU","ASN"))
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

####*7217 genes.*

#\ 

###**8. Over 80% of genes are called differentially expressed between ethnic groups. However, due to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the “major effect” in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005. How many genes have q-values < 0.05?**

```{r, eval = FALSE}
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
index = which(ethnicity%in% c("CEU","ASN") & year=="05")
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

####*560 genes.*

#\ 

###**9. To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3) How many genes have q-values < 0.05?**

```{r, eval = FALSE}
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
indexASN = which(ethnicity%in% c("ASN") & year=="05")
indexCEU = which(ethnicity%in% c("CEU") & year=="02")
set.seed(3)
indexCEU <- sample(indexCEU,3)
index <- c(indexASN,indexCEU)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

####*3695 genes.*

#\ 

#Batch Effects Exercise List 3: 

###**1. Using the functions rowttests and qvalue compare the two groups. Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%. How many gene have q-values less than 0.1?** 

```{r, eval = FALSE}
sex = factor (format(sampleInfo$group)) 
month = factor (format(sampleInfo$date,"%m")) 
table(sampleInfo$group, month)
ttest <- rowttests(geneExpression, sex)
pval <- ttest$p.val
qval <- qvalue(pval)
sum(qval$qvalues < 0.1)
```

####*59 genes.*

#\ 

###**2. Note that sampleInfo$group here presents males and females. Thus, we expect differences to be in on chrY and, for genes that escape inactivation, chrX. We do not expect many autosomal genes to be different between males and females. This gives us an opportunity to evaluate false and true positives with experimental data. For example, we evaluate results using the proportion genes of the list that are on chrX or chrY. For the list calculated above, what proportion of this list is on chrX or chrY?** 

```{r, eval = FALSE}
chrXorY <- geneAnnotation$CHR[qval$qvalues < 0.1]
mean(chrXorY == "chrX" | chrXorY == "chrY")
```

####*0.3389831.*

#\ 

###**3. We can also check how many of the chromosomes X and Y genes we detected as different. How many are on Y?** 

```{r, eval = FALSE}
chrXorY <- geneAnnotation$CHR[qval$qvalues < 0.1]
mean(chrXorY == "chrY")
```

####*#Proportion of 0.1355932.*

#\ 

###**4. Now for the autosomal genes (not on chrX and chrY) for which q-value < 0.1, perform a t-test comparing samples processed in June to those processed in October. What proportion of these have p-values <0.05?**

```{r, eval = FALSE}
filter <- geneAnnotation[geneAnnotation$CHR != "chrX" & geneAnnotation$CHR != "chrY" & qval$qvalues < 0.1, ]
autosomal <- geneExpression[filter$PROBEID,]
month = factor (format(sampleInfo$date,"%m"))
ttest <- rowttests(autosomal, month)
mean(ttest$p.val < 0.05)
```

####*0.8717949.*

#\ 

###**5. The above result shows that the great majority of the autosomal genes show differences due to processing data. This provides further evidence that confounding is resulting in false positives. So we are going to try to model the month effect to better estimate the sex effect. We are going to use a linear model: Which of the following creates the appropriate design matrix?** 

####**A) X = model.matrix(~sex+ethnicity).**

####**B) X = cbind(sex,as.numeric(month)).**

####**C) It can’t be done with one line.** 

####**D) X = model.matrix(~sex+month).** 

```{r, eval = FALSE}
model.matrix(~sex+month)
```

####*D.*

#\ 

###**6. How many of the q-values for the group comparison are now <0.1? Note the big drop from what we obtained without the correction.** 

```{r, eval = FALSE}
pvals <- vector()
for(i in 1:nrow(geneExpression)){
        X = model.matrix(~sex+month)
        y = geneExpression[i,]
        fit = lm(y ~ X - 1)
        qv <- summary(fit)
        pvals[i] <- qv$coefficients[2,4]
}
qvals <- qvalue(pvals)
sum(qvals$qvalues < 0.1)
```

####*Dropped to 17.*

#\ 

###**7. With this new list, what proportion of these are chrX and chrY? Notice the big improvement.**

```{r, eval = FALSE}
chrXorY <- geneAnnotation$CHR[qvals$qvalues < 0.1]
mean(chrXorY == "chrY")
mean(chrXorY == "chrX")
```

####*proprtion of 0.3529412 on Y. Proportion of 0.5294118 on X.*

#\ 

###**8. How many on Y or X?** 

```{r, eval = FALSE}
chrXorY <- geneAnnotation$CHR[qvals$qvalues < 0.1]
mean(chrXorY == "chrX" | chrXorY == "chrY")
```

####*From 0.33 all the way to 0.8823529.*

#\ 

###**9. Now from the linear model above, extract the p-values related to the coefficient representing the October versus June differences using the same linear model. How many of the q-values for the month comparison are now <0.1? This approach is basically the approach implemented by Combat.**

```{r, eval = FALSE}
ComBatvals <- vector()
for(i in 1:nrow(geneExpression)){
        X = model.matrix(~ sex + month)
        y = geneExpression[i,]
        fit = lm(y ~ X - 1)
        qv <- summary(fit)
        ComBatvals[i] <- qv$coefficients[3, 4]
}
ComBatvals <- qvalue(ComBatvals)
sum(ComBatvals$qvalues < 0.1)
```

####*3170 q-vals.*

#\ 

#Batch Effects Exercise List 4:

###**1. Suppose you want to make an MA plot of the first two samples y = geneExpression[,1:2]. Which of the following projections gives us the projection of y so that column2 versus column 1 is an MA plot?** 

####**A. y(1/sqrt(2) 1/sqrt(2) 1sqrt(2) -1/sqrt(2)**  

####**B. y(1 1 1 -1)** 

####**C. (1 1 1 -1)y** 

####**D. (1 1 1 -1)y'** 

```{r, eval = FALSE}
y = geneExpression[,1:2]
```

####*B.*

#\ 

###**2. Say Y is M ×N, in the SVD Y = UDV' which of the following is not correct?**

####**A) DV' are the new coordinates for the projection U'Y.**

####**B) UD are the new coordinates for the projection Y V.**

####**C) D are the coordinates of the projection U'Y.** 

####**D) U'Y is a projection from an N-dimensional to M-dimensional subspace.**

####*C.*

#\ 

###**3. Define:**
```{r, eval = FALSE}
y = geneExpression - rowMeans(geneExpression)
```

###**Compute and plot an image of the correlation for each sample. Make two image plots of these correlations. In the first one, plot the correlation as image. In the second, order the samples by date and then plot an image of the correlation. The only difference in these plots is the order in which the samples are plotted. Based on these plots, which of the following you would say is true?** 

####**A) The samples appear to be completely independent of each other.** 

####**B) Sex seems to be creating structures as evidenced by the two cluster of highly correlated samples.**

####**C) There appear to be only two factors completely driven by month.** 

####**D) The fact that in the plot ordered by month we see two groups mainly driven by month, and within these we see subgroups driven by date, seems to suggest date more than month per se are the hidden factors.**

```{r, eval = FALSE}
mypar(1,2)
n=ncol(y)
cors=cor(y)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
o <- order(sampleInfo$date)
yo<-y[,o]
cors=cor(yo)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
```

####*C. Only two factors seem to be driven by month and ordered by sex.*

#\ 

###**4. Based on the correlation plots above, we could argue that there are at least two hidden factors. Using PCA estimate these two factors. Specifically, apply the svd to y and use the first two PCs as estimates. Which command gives us these estimates?**

####**A) pcs = svd(y)$v[1:2,]** 

####**B) pcs = svd(y)$v[,1:2]** 

####**C) pcs = svd(y)$u[,1:2]** 

####**D) pcs = svd(y)$d[1:2]** 

```{r, eval = FALSE}
Est <- svd(y)$v[,1:2]
rownames(Est)<-colnames(y)
round(Est,2)
fit = svd(y)$u[,1:2]%*% (svd(y)$d[1:2]*t(Est))
var(as.vector(fit))/var(as.vector(y))
```

####*B.*

#\ 

###**5. Plot each of the estimated factors ordered by date. Use color to denote month. The first factor is clearly related to date. Which of the following appear to be most different according to this factor?** 

####**A) June 23 and June 27** 

####**B) Oct 07 and Oct 28**

####**C) June 10 and June 23** 

####**D) June 15 and June 24** 

```{r, eval = FALSE}
month=format(sampleInfo$date, "%m")
pcs = svd(y)$v[,1:2]
o = order(sampleInfo$date)
cols = as.numeric(month)[o]
mypar(2,1)
for(i in 1:2){
  plot(pcs[o,i],col=cols,xaxt="n",xlab="")
  label = gsub("2005-","",sampleInfo$date[o])
  axis(1,1:ncol(y),label,las=2)
}
```

####*A.*

#\ 

###**6. Use the svd function to obtain the principal components (PCs) for our detrended gene expression data y. How many PCs explain more than 10% of the variability?** 

```{r, eval = FALSE}
s = svd(y)
varexplained = s$d^2/ sum(s$d^2)
plot(varexplained)
sum(varexplained>0.10)
```

####*PC 1 & PC 2 explain more than 10% of variability.*

#\ 

###**7. Which PC most correlates (negative or positive correlation) with month?** 

```{r, eval = FALSE}
month = factor( format(sampleInfo$date,"%m"))
cors = cor( as.numeric(month),svd(y)$v)
plot(t(cors))
which.max(abs(cors))
```

####*PC 1, PC 2.*

#\ 

###**8. What is this correlation (in absolute value)?** 

```{r, eval = FALSE}
max(abs(cors))
```

####*0.8297915*

#\ 

###**9. Which PC most correlates (negative or positive correlation) with sex?** 

```{r, eval = FALSE}
sex = sampleInfo$group
cors = cor( as.numeric(sex),svd(y)$v)
plot(t(cors))
which.max(abs(cors))
```

####*PC 1.*

#\ 

###**10. What is this correlation (in absolute value)?**

```{r, eval = FALSE}
max(abs(cors))
```

####*0.6236858.*

#\  

###**11. Now instead of using month, which we have shown does not quite describe the batch, add the two estimated factors s$v[,1:2] to the linear model we used above. Apply this model to each gene and compute q-values for the sex difference. How many q-values < 0.1 for the sex comparison?**

```{r, eval = FALSE}
X = model.matrix(~sex+month)
pvals <- t( sapply(1:nrow(geneExpression),function(j){
  y <- geneExpression[j,]
  fit <- lm(y~X-1)
  summary(fit)$coef[3,4]
  }))
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

####*3170 q-values.*

#\ 

###**12. What proportion of the genes are on chromosomes X and Y?**

```{r, eval = FALSE}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

####*0.04763407.*